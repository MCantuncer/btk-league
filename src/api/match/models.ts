import { Field, InputType } from 'type-graphql';
import { ValidateObjectId } from '../../helpers/validators';
import { User } from '../../models/user/entity';

@InputType()
export class MatchSearchInput {
  @Field(() => String)
  @ValidateObjectId(User)
  home?: string;

  @Field(() => String)
  @ValidateObjectId(User)
  visitor?: string;

  @Field(() => Date)
  matchDate?: Date;
}
