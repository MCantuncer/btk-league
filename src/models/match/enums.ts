import { registerEnumType } from 'type-graphql';

export enum WinnerEnum {
  HOME = 1,
  VISITOR = 2,
}

registerEnumType(WinnerEnum, {
  name: 'WinnerEnum',
  description: 'Winner Enums',
});
