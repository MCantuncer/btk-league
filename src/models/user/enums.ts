import { registerEnumType } from 'type-graphql';

export enum UserTypeEnum {
  COACH = 1,
  PLAYER = 2,
  SPECTATOR = 3,
}

registerEnumType(UserTypeEnum, {
  name: 'UserTypeEnum',
  description: 'User Types',
});
