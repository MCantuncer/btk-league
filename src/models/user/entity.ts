import { Field, Int, ObjectType } from 'type-graphql';
import { getModelForClass, prop } from '@typegoose/typegoose';
import { UserTypeEnum } from './enums';
import { EntityWithIdAndDates } from '../base';

@ObjectType()
export class User extends EntityWithIdAndDates {
  @prop()
  @Field(() => UserTypeEnum)
  userType: UserTypeEnum;

  @prop()
  @Field(() => String)
  firstName: string;

  @prop()
  @Field(() => String)
  lastName: string;

  @prop()
  @Field(() => String)
  email: string;

  @prop()
  @Field(() => String)
  password: string;

  @prop()
  @Field(() => Int, { nullable: true })
  standing?: number | null;
}

export const UserModel = getModelForClass(User, {
  schemaOptions: { collection: 'users' },
});
