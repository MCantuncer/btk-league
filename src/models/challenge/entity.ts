import { Field, ObjectType } from 'type-graphql';
import { EntityWithIdAndDates } from '../base';
import { getModelForClass, prop, Ref } from '@typegoose/typegoose';
import { User } from '../user/entity';

@ObjectType()
export class Challenge extends EntityWithIdAndDates {
  @prop({ ref: User, type: String })
  @Field(() => String)
  challenger: Ref<User>;

  @prop({ ref: User, type: String })
  @Field(() => String)
  challenged: Ref<User>;

  @prop()
  @Field(() => Boolean, { nullable: true })
  accepted?: boolean | null;
}

export const ChallengeModel = getModelForClass(Challenge, {
  schemaOptions: { collection: 'challenges' },
});
