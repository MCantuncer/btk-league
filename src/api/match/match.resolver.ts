import { Match, MatchModel } from '../../models/match/entity';
import { Arg, Args, Authorized, Mutation, Query, Resolver } from 'type-graphql';
import { DocumentQuery } from 'mongoose';
import { MatchArgs, MatchPaginationResponse } from './models';
import { ModelUtils } from '../../helpers/database';
import { DocumentType } from '@typegoose/typegoose';
import { MatchInput } from '../../models/match/input';
import { MatchManager } from '../../models/match/manager';
import { PaginationInput } from '../../graphql/filters';
import { IPaginationResponse } from '../../graphql/pagination';

@Resolver(() => Match)
export default class {
  @Authorized()
  @Query(() => MatchPaginationResponse)
  async listMatches(@Args(() => MatchArgs) args: MatchArgs): Promise<MatchPaginationResponse> {
    let query: DocumentQuery<DocumentType<Match>[], DocumentType<Match>, {}> = (
      await ModelUtils.findAndFilter(MatchModel, args)
    )();

    const paginationInput = args['pagination'] || new PaginationInput();
    return (await ModelUtils.applyPagination(
      query,
      paginationInput,
      MatchPaginationResponse,
    )) as IPaginationResponse<Match>;
  }

  @Authorized()
  @Mutation(() => Match)
  async upsertMatch(@Arg('input', () => MatchInput) input: MatchInput): Promise<Match | null> {
    return await MatchManager.upsertMatch(input);
  }
}
