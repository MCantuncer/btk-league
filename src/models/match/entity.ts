import { Field, GraphQLTimestamp, Int, ObjectType } from 'type-graphql';
import { EntityWithIdAndDates } from '../base';
import { getModelForClass, modelOptions, prop, Ref } from '@typegoose/typegoose';
import { User } from '../user/entity';
import { WinnerEnum } from './enums';

@ObjectType()
@modelOptions({ schemaOptions: { _id: false } })
export class SetResult {
  @prop()
  @Field(() => Int)
  homeGame: number;

  @prop()
  @Field(() => Int)
  visitorGame: number;

  @prop()
  @Field(() => WinnerEnum, { nullable: true })
  winner?: WinnerEnum | null;
}

@ObjectType()
export class Match extends EntityWithIdAndDates {
  @prop({ ref: User, type: String })
  @Field(() => String)
  home: Ref<User>;

  @prop({ ref: User, type: String })
  @Field(() => String)
  visitor: Ref<User>;

  @prop()
  @Field(() => [SetResult], { nullable: true })
  result?: SetResult[] | null;

  @prop()
  @Field(() => WinnerEnum, { nullable: true })
  matchWinner?: WinnerEnum | null;

  @prop()
  @Field(() => GraphQLTimestamp, { nullable: true })
  matchDate?: Date | null;
}

export const MatchModel = getModelForClass(Match, {
  schemaOptions: { collection: 'matches' },
});
