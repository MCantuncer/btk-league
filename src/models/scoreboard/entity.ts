import { Field, Int, ObjectType } from 'type-graphql';
import { prop, Ref } from '@typegoose/typegoose';
import { User } from '../user/entity';

@ObjectType()
export class ScoreBoard {
  @prop({ ref: User, type: String })
  @Field(() => String)
  user: Ref<User>;

  @prop()
  @Field(() => Int)
  points: number;
}
