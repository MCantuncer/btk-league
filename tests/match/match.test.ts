import { testMutate, testQuery } from '../../jest.setup';
import { UserInputs, UserMutations } from '../user/globals';
import { MatchInputs, MatchMutations, MatchQueries } from './globals';
import { Match } from '../../src/models/match/entity';
import { WinnerEnum } from '../../src/models/match/enums';
import moment from 'moment';

describe('Match Tests', () => {
  it('Create a Match', async () => {
    let graphqlResponse = await testMutate(UserMutations.Instance.REGISTER, {
      input: { ...UserInputs.Instance().REGISTER_INPUT_1, email: 'tuncer@btk.com' },
    });

    expect(graphqlResponse.errors).toBeUndefined();

    graphqlResponse = await testMutate(UserMutations.Instance.REGISTER, {
      input: { ...UserInputs.Instance().REGISTER_INPUT_1, email: 'can@btk.com', id: UserInputs.CREATE_ID_2 },
    });

    expect(graphqlResponse.errors).toBeUndefined();

    graphqlResponse = await testMutate(MatchMutations.Instance.UPSERT_MATCH, {
      input: MatchInputs.Instance().UPSERT_MATCH_INPUT_1,
    });

    expect(graphqlResponse.errors).toBeUndefined();

    const response = (graphqlResponse.data || {}) as { upsertMatch: Match };

    expect(response.upsertMatch.home).toEqual(UserInputs.CREATE_ID_1);
    expect(response.upsertMatch.visitor).toEqual(UserInputs.CREATE_ID_2);

    expect(response.upsertMatch.result).toBeNull();
    expect(response.upsertMatch.matchWinner).toBeNull();

    response.upsertMatch.matchDate = null;

    expect(response).toMatchSnapshot();
  });
  it('Update a Match', async () => {
    const graphqlResponse = await testMutate(MatchMutations.Instance.UPSERT_MATCH, {
      input: {
        ...MatchInputs.Instance().UPSERT_MATCH_INPUT_1,
        result: [
          { homeGame: 3, visitorGame: 6, winner: WinnerEnum[WinnerEnum.VISITOR] },
          { homeGame: 3, visitorGame: 6, winner: WinnerEnum[WinnerEnum.VISITOR] },
        ],
        matchWinner: WinnerEnum[WinnerEnum.VISITOR],
      },
    });

    expect(graphqlResponse.errors).toBeUndefined();

    const response = (graphqlResponse.data || {}) as { upsertMatch: Match };

    expect(response.upsertMatch.matchWinner).toEqual(WinnerEnum[WinnerEnum.VISITOR]);
    expect(response.upsertMatch.result).toHaveLength(2);

    response.upsertMatch.matchDate = null;

    expect(response).toMatchSnapshot();
  });
  it('Create new Match', async () => {
    const graphqlResponse = await testMutate(MatchMutations.Instance.UPSERT_MATCH, {
      input: MatchInputs.Instance().UPSERT_MATCH_INPUT_2,
    });

    expect(graphqlResponse.errors).toBeUndefined();

    const response = (graphqlResponse.data || {}) as { upsertMatch: Match };

    expect(response.upsertMatch.home).toEqual(UserInputs.CREATE_ID_2);
    expect(response.upsertMatch.visitor).toEqual(UserInputs.CREATE_ID_1);

    expect(response.upsertMatch.result).toBeNull();
    expect(response.upsertMatch.matchWinner).toBeNull();

    response.upsertMatch.matchDate = null;

    expect(response).toMatchSnapshot();
  });
  it('List Matches', async () => {
    const graphqlResponse = await testQuery(MatchQueries.Instance.LIST_MATCHES);

    expect(graphqlResponse.errors).toBeUndefined();
  });
  it('Filter Matches by Home Player', async () => {
    const graphqlResponse = await testQuery(MatchQueries.Instance.LIST_MATCHES, {
      home: { eq: UserInputs.CREATE_ID_1 },
    });

    expect(graphqlResponse.errors).toBeUndefined();

    const response = (graphqlResponse.data || {}) as { listMatches: { data: Match[] } };

    removeMatchDateFromMatchList(response.listMatches.data);

    expect(response.listMatches.data).toHaveLength(1);
    expect(response.listMatches.data[0].home).toEqual(UserInputs.CREATE_ID_1);

    expect(response).toMatchSnapshot();
  });
  it('Filter Matches by Match Date', async () => {
    const graphqlResponse = await testQuery(MatchQueries.Instance.LIST_MATCHES, {
      matchDate: { gte: parseInt(moment(new Date()).add('-10', 'minutes').format('x')) },
    });

    expect(graphqlResponse.errors).toBeUndefined();

    const response = (graphqlResponse.data || {}) as { listMatches: { data: Match[] } };

    removeMatchDateFromMatchList(response.listMatches.data);

    expect(response.listMatches.data).toHaveLength(1);
    expect(response.listMatches.data[0].id).toEqual(MatchInputs.CREATE_ID_2);

    expect(response).toMatchSnapshot();
  });
});

const removeMatchDateFromMatchList = (matchList: Match[]) => {
  for (const match of matchList) {
    match.matchDate = null;
  }
};
