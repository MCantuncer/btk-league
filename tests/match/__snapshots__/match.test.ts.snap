// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match Tests Create a Match 1`] = `
Object {
  "upsertMatch": Object {
    "deleted": false,
    "home": "0bbdcc9f-a146-4d1b-a2bc-65257dadbd48",
    "id": "5bacbea0-8126-41bb-a9da-e20d86398e69",
    "matchDate": null,
    "matchWinner": null,
    "result": null,
    "visitor": "05837b13-6c10-45f8-9ed5-a88fa6ef9ff2",
  },
}
`;

exports[`Match Tests Create new Match 1`] = `
Object {
  "upsertMatch": Object {
    "deleted": false,
    "home": "05837b13-6c10-45f8-9ed5-a88fa6ef9ff2",
    "id": "ce6429be-8660-4282-b6f6-4f07c4b434be",
    "matchDate": null,
    "matchWinner": null,
    "result": null,
    "visitor": "0bbdcc9f-a146-4d1b-a2bc-65257dadbd48",
  },
}
`;

exports[`Match Tests Filter Matches by Home Player 1`] = `
Object {
  "listMatches": Object {
    "count": 1,
    "data": Array [
      Object {
        "deleted": false,
        "home": "0bbdcc9f-a146-4d1b-a2bc-65257dadbd48",
        "id": "5bacbea0-8126-41bb-a9da-e20d86398e69",
        "matchDate": null,
        "matchWinner": "VISITOR",
        "result": Array [
          Object {
            "homeGame": 3,
            "visitorGame": 6,
            "winner": "VISITOR",
          },
          Object {
            "homeGame": 3,
            "visitorGame": 6,
            "winner": "VISITOR",
          },
        ],
        "visitor": "05837b13-6c10-45f8-9ed5-a88fa6ef9ff2",
      },
    ],
    "hasNext": false,
    "limit": 50,
    "page": 1,
  },
}
`;

exports[`Match Tests Filter Matches by Match Date 1`] = `
Object {
  "listMatches": Object {
    "count": 1,
    "data": Array [
      Object {
        "deleted": false,
        "home": "05837b13-6c10-45f8-9ed5-a88fa6ef9ff2",
        "id": "ce6429be-8660-4282-b6f6-4f07c4b434be",
        "matchDate": null,
        "matchWinner": null,
        "result": null,
        "visitor": "0bbdcc9f-a146-4d1b-a2bc-65257dadbd48",
      },
    ],
    "hasNext": false,
    "limit": 50,
    "page": 1,
  },
}
`;

exports[`Match Tests Update a Match 1`] = `
Object {
  "upsertMatch": Object {
    "deleted": false,
    "home": "0bbdcc9f-a146-4d1b-a2bc-65257dadbd48",
    "id": "5bacbea0-8126-41bb-a9da-e20d86398e69",
    "matchDate": null,
    "matchWinner": "VISITOR",
    "result": Array [
      Object {
        "homeGame": 3,
        "visitorGame": 6,
        "winner": "VISITOR",
      },
      Object {
        "homeGame": 3,
        "visitorGame": 6,
        "winner": "VISITOR",
      },
    ],
    "visitor": "05837b13-6c10-45f8-9ed5-a88fa6ef9ff2",
  },
}
`;
